### PR TITLE
core: first step towards a pidfd focused future

### PR DESCRIPTION
**Pull request by poettering**
This is the first step towards putting pidfds front and center for service process tracking. It doesn't move all process operations done by PID 1 over to it, but at least the basic tracking is now via pidfd. Further infra should be moved over too, but can happen in separate steps, already to make this PR not ridiculously large.

### CODE REVIEW

File: src/basic/meson.build [hunk 1]
```diff
@@ -65,6 +65,8 @@ basic_sources = files(
         'path-lookup.c',
         'path-util.c',
         'percent-util.c',
+        'pidref.c',
+        'pidref.h',
```
Comment (YHNdnzj) [inline src/basic/meson.build]: Headers are handled automatically by meson. You don't need to add them here.

File: src/basic/pidref.c [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+                if (!ERRNO_IS_RESOURCE(errno))
+                        return -errno;
+
+                /* Graceful fallback if we are out of fds */
+                r = pidfd_get_pid(fd, &pid);
+                if (r < 0)
+                        return r;
+
+                return pidref_set_pid(pidref, pid);
```
Comment (YHNdnzj) [inline src/basic/pidref.c]: This seems confusing, as in the fallback path (`pidref_set_pid`) we try to open a new pidfd again. Maybe make `pidref_set_pid` only touch PidRef.pid, and introduce `pidref_set_both`?

File: src/basic/pidref.c [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        }
+
+        r = pidref_set_pidfd_take(pidref, fd_copy);
+        if (r < 0)
+                return r;
+
+        TAKE_FD(fd_copy);
```
Comment (YHNdnzj) [inline src/basic/pidref.c]: Introduce the `_consume` variant?

File: src/basic/pidref.c [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+PidRef* pidref_done(PidRef *pidref) {
+        if (!pidref)
+                return NULL;
+
+        *pidref = (PidRef) {
+                .fd = safe_close(pidref->fd),
+        };
+
+        return NULL;
```
Comment (YHNdnzj) [inline src/basic/pidref.c]: For structs that are usually allocated on the stack, we usually just make the `_done` destructor a `void` function?

File: src/basic/pidref.c [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        *pidref = (PidRef) {
+                .fd = fd,
+                .pid = pid,
+        };
+
+        return 0;
+}
+
+PidRef* pidref_done(PidRef *pidref) {
+        if (!pidref)
```
Comment (YHNdnzj) [inline src/basic/pidref.c]: (And change this to an assert)

File: src/basic/pidref.h [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+int pidref_set_pidfd_take(PidRef *pidref, int fd); /* takes ownership of the passed pidfd */
+
+PidRef* pidref_done(PidRef *pidref);
+
+int pidref_kill(PidRef *pidref, int sig);
+int pidref_kill_and_sigcont(PidRef *pidref, int sig);
+
+DEFINE_TRIVIAL_CLEANUP_FUNC(PidRef*, pidref_done);
+
+#define TAKE_PIDREF(p) ({ PidRef t = (p); (p) = PIDREF_INIT; t; })
```
Comment (YHNdnzj) [inline src/basic/pidref.h]: `TAKE_GENERIC(p, PidRef, PIDREF_INIT)`

File: src/basic/pidref.h [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+/* An embeddable structure carrying a reference to a process. Supposed to be used when tracking processes continously. */
+typedef struct PidRef {
+        pid_t pid; /* always valid */
+        int fd;    /* only valid if pidfd are available in the kernel, and we manage to get an fd */
+} PidRef;
+
+#define PIDREF_NULL (PidRef) { .fd = -EBADF }
+
+static inline bool pidref_is_set(const PidRef *pidref) {
+        return pidref && pidref->pid > 0;
```
Comment (YHNdnzj) [inline src/basic/pidref.h:15]: We always embed the structure, so this can be `ASSERT_PTR(pidref)->pid ...` (Don't really matter)

**Review by bluca — APPROVED** [review]
nice stuff

File: src/basic/pidref.h (hunk 7 above)
Comment (bluca) [reply src/basic/pidref.h:15]: I might need to use it as-is in the sd-exec PR, so let's keep it like this for now

**Pull request merged**